### PR TITLE
[bug] Fix Pytorch profiler with emit_nvtx

### LINKED ISSUE #6153: PyTorchProfiler crashes when emit_nvtx=True
## 🐛 Bug

When training with PyTorchProfiler(emit_nvtx=True), the training stops with the following error : 
`AttributeError: 'emit_nvtx' object has no attribute 'function_events'`

## Please reproduce using the BoringModel

-> https://colab.research.google.com/drive/1cqMxMgDVgltluaYZAkn9d2srSDmOu-1p?usp=sharing

### To Reproduce

Use following [**BoringModel**](**https://colab.research.google.com/drive/1cqMxMgDVgltluaYZAkn9d2srSDmOu-1p?usp=sharing**) and post here

### Additional context

Additional docs issue : 
It says here (https://pytorch-lightning.readthedocs.io/en/1.2.0/advanced/profiler.html) that you have to use `nvprof` to collect the profiler traces, but it is no longer supported for devices with compute capability 8.0 and higher.
I think that an exemple using nsight-compute would be great ( I can't test it right now because of the nvtx issue )

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolve a bug with PyTorch Profiler and `emit_nvtx`.

```
nvprof --profile-from-start off -o trace_name.prof -- pytest tests/trainer/test_trainer.py::test_pytorch_profiler_nested_emit_nvtx

python -c "import torch;print(torch.autograd.profiler.load_nvprof('/home/ubuntu/pytorch-lightning/trace_name.prof'))"
```

Screenshot 2021-03-01 at 13 56 36

Fixes #6153

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/test_trainer.py [hunk 1]
```diff
@@ -1664,6 +1664,25 @@ def test_pytorch_profiler_nested(tmpdir):
         assert pa[n] == expected_[n]
 
 
+@pytest.mark.skipif(True, reason="This tests need to be run with nvprof")
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1667]:
```suggestion
@pytest.mark.skip(reason="This tests need to be run with nvprof")
```

File: pytorch_lightning/profiler/profilers.py [hunk 2]
```diff
@@ -466,15 +466,22 @@ def _create_profiler(self, action_name, profiler, enter=True):
...
         self.profiler = pr
+        return pr
 
     def _stop(self, action_name: str) -> None:
         if self.profiler is None:
             return
 
         self.profiler.__exit__(exc_type=None, exc_val=None, exc_tb=None)
 
+        if isinstance(self.profiler, torch.autograd.profiler.emit_nvtx):
```
Comment (carmocca) [inline pytorch_lightning/profiler/profilers.py:481]: so running __enter__ replaces the profiler with `torch.autograd.profiler.emit_nvtx`?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/profiler/profilers.py [hunk 3]
```diff
@@ -466,15 +466,22 @@ def _create_profiler(self, action_name, profiler, enter=True):
         profiler_args = {k: v for k, v in vars(self).items() if k in init_args}
         pr = profiler(**profiler_args)
         if enter:
-            pr = pr.__enter__()
+            out_pr = pr.__enter__()
+            if out_pr is not None:
+                pr = out_pr
         self.profiler = pr
+        return pr
```
Comment (Borda) [inline pytorch_lightning/profiler/profilers.py:473]:
```suggestion
        return self.profiler
```
why do you need this return?

File: pytorch_lightning/profiler/profilers.py [hunk 4]
```diff
@@ -466,15 +466,22 @@ def _create_profiler(self, action_name, profiler, enter=True):
...
 
     def _stop(self, action_name: str) -> None:
         if self.profiler is None:
             return
 
         self.profiler.__exit__(exc_type=None, exc_val=None, exc_tb=None)
 
+        if isinstance(self.profiler, torch.autograd.profiler.emit_nvtx):
+            self._profiler.__exit__(None, None, None)
```
Comment (Borda) [inline pytorch_lightning/profiler/profilers.py:482]: it is a bit confusing, what is diff between `self.profiler` and `self._profiler`

File: tests/trainer/test_trainer.py [hunk 5]
```diff
@@ -1664,6 +1664,25 @@ def test_pytorch_profiler_nested(tmpdir):
...
 
+@pytest.mark.skip(reason="This tests need to be run with nvprof")
+@pytest.mark.skipif(torch.cuda.device_count() < 1, reason="test requires a GPU machine")
+def test_pytorch_profiler_nested_emit_nvtx(tmpdir):
+    """
+    How to run this test.
+    nvprof --profile-from-start off -o trace_name.prof -- pytest {}:test_pytorch_profiler_nested_emit_nvtx
+    python -c "import torch;print(torch.autograd.profiler.load_nvprof('{}/trace_name.prof'))"
+    """
+    profiler = PyTorchProfiler(use_cuda=True, emit_nvtx=True)
```
Comment (Borda) [inline tests/trainer/test_trainer.py:1675]: this is just about that it does not crash, how you know that all is fine?

File: pytorch_lightning/profiler/profilers.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/profiler/profilers.py:481]: Not really, it needs to open 2 profilers to work ...

File: pytorch_lightning/profiler/profilers.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/profiler/profilers.py:482]: Renamed to parent_profiler.

File: tests/trainer/test_trainer.py (hunk 5 above)
Comment (tchaton) [reply tests/trainer/test_trainer.py:1675]: Great question ! I wanted to see with you if we should run this one in the CI.

Comment (Borda) [reply tests/trainer/test_trainer.py:1580]: what you mean to run in CI, we shall test all our code :]

File: pytorch_lightning/profiler/profilers.py [hunk 6]
```diff
@@ -291,3 +294,270 @@ def __del__(self):
         """Close profiler's stream."""
         if self.output_file:
             self.output_file.close()
+
+
+class PyTorchProfiler(BaseProfiler):
```
Comment (Borda) [inline pytorch_lightning/profiler/profilers.py:299]: all this class was moved to profiler/pytorch.py

**Review by Borda — DISMISSED** [review]

File: tests/special_tests.sh [hunk 7]
```diff
@@ -32,3 +32,6 @@ python ${DEFAULTS} tests/trainer/test_trainer.py::test_pytorch_profiler_trainer_
 python ${DEFAULTS} tests/models/test_hooks.py::test_transfer_batch_hook_ddp
 python ${DEFAULTS} tests/trainer/test_data_loading.py::test_replace_distrubuted_sampler_custom_dataloader_custom_batch_sampler
 python ${DEFAULTS} tests/trainer/optimization/test_manual_optimization.py::test_step_with_optimizer_closure_with_different_frequencies_ddp_with_toggle_model
+
+nvprof --profile-from-start off -o trace_name.prof -- python ${DEFAULTS} tests/trainer/test_trainer.py:test_pytorch_profiler_nested_emit_nvtx
```
Comment (carmocca) [inline tests/special_tests.sh:36]: Does this run the test even if it's marked with `skip`?

File: tests/trainer/test_trainer.py [hunk 8]
```diff
@@ -1568,6 +1568,26 @@ def test_pytorch_profiler_nested(tmpdir):
         assert pa[n] == expected_[n]
 
 
+@pytest.mark.skip(reason="This tests need to be run with nvprof")
+@pytest.mark.skipif(torch.cuda.device_count() < 1, reason="test requires a GPU machine")
+@pytest.mark.skipif(
+    not os.getenv("PL_RUNNING_SPECIAL_TESTS", '0') == '1', reason="test should be run outside of pytest"
+)
```
Comment (carmocca) [inline tests/trainer/test_trainer.py:1575]: Use `RunIf`

File: pytorch_lightning/profiler/pytorch.py [hunk 9]
```diff
@@ -118,13 +118,6 @@ def __init__(
 
             local_rank: When running in distributed setting, local_rank is used for each process
                 to write to their own file if `output_fname` is provided.
-
-        Raises:
-            MisconfigurationException:
-                If arg ``sort_by_key`` is not present in ``AVAILABLE_SORT_KEYS``, or
-                if log file is not a ``.txt`` file.
-            ValueError:
-                If you attempt to stop recording an action which was never started.
```
Comment (carmocca) [inline pytorch_lightning/profiler/pytorch.py:127]: Why are these gone?

File: tests/trainer/test_trainer.py (hunk 8 above)
Comment (tchaton) [reply tests/trainer/test_trainer.py:1557]: Yes !

**Review by justusschock — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**